accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,185 @@
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityTagsCollection? tags)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
+    }
+```
+
+### ActivityLink
+
+```C#
+    // Already exposed APIs
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
+    }
+```
+
+## Automatic Trace Id  generation in case of null parent
+
+Sampler like probabilistic sampler can depend on the trace id for the sampling decision. Usually before creating any Activity using ActivitySource, the `ActivityListener.GetRequestedDataUsingContext` callback will get called which usually get implemented by the samplers. The parent context is sent as a parameter to this callback and the sampler gets the trace id from this parent context to perform the sampling action. When there is no parent, we send the default context which is nulling the trace id. The samplers depending on trace id are not going to be able to do the sampling at that time. The samplers even cannot generate and use a trace id because it cannot inform the Activity creation process to use the generated trace Id when creating the Activity object.
+There are a couple of proposals how to address this issue in the .NET runtime. The idea is the runtime need to generate the trace Id and send it to the samplers. If the sampler using the passed trace id and decide to sample in this case, the Activity which will get created because of that sampler decision will be using the generated trace Id in its context.
+
+Solution options:
+
+- Before calling `ActivityListener.GetRequestedDataUsingContext`, there will be a check if we don't have a parent, then we generate a random trace Id and use it in the context sent to the callback. If decided to sample in, we all use this trace Id during the Activity object creation. This option doesn't need any API changes and will be fully handled inside the implementation.
+- Add a new property to the `ActivityCreationOptions` struct (e.g. `NullParentTraceId`) which will carry the trace id that can be used by the sampler when detecting no parent context. obviously this will need API change.
+
+I am inclining to the first solution for the following reasons:
```

Comment:
Sounds fine to me, lets make sure to document carefully : ) The current property name says that the ActivityContext represents the parent and its a somewhat creative interpretation to use the TraceId portion of it as a newly generated random value instead.

Reply:
Sounds good. This change will not need API change. if all agree, I can start implementing this part from now.

Reply:
Yes. this must be documented clearly to avoid confusion.

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#shouldsample

OT spec for sampling makes this separate - i.e parentcontext and traceid are passed separately. Would it make sense to do that here? So that there would be clear expectation what the traceid would be - it'd be the same traceid as parentcontext.traceid. if parent is null, then traceid is the id of the "to be created" activity.

Reply:
The spec is saying `TraceId of the Span to be created. It can be different from the TraceId in the SpanContext. Typically in situations when the Span to be created starts a new Trace.`. do we have any real scenario the samplers will need to get 2 trace Ids (one from the parent and one for the span to be created with)? how this will be useful to the samplers? I am expecting whoever calling `ActivitySource.StartActivity` is going to ensure passing the trace id in the parent context to have the Activity object created with this trace Id. 

I am expecting the text in this part of the document is going to change to accommodate the points that @noahfalk raised. I'll expect there will be some API change here.